client/python/client_cache.py

```diff
@@ -0,0 +1,135 @@
+# Copyright 2014 Treode, Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import string
+from urllib3 import HTTPConnectionPool
+import json
+
+from cache_map import *
+from http_facade import *
+from tx_clock import *
+from stale_exception import *
+
+"""
+Treode Python client cache implementation
+"""
+class ClientCache(object):
+
+    MAX_CACHE_SIZE = 1000
+
+    def __init__(self, server, port=80, max_age=None, no_cache=False):
+        self.server = server
+        self.port = port
+        self.max_age = max_age
+        self.no_cache = no_cache
+
+        # Initialize connection to the DB server
+        http_facade = HTTPFacade(server, port)
+        self.http_facade = http_facade
+
+        # Initialize the cache value store.
+        self.cache_map = CacheMap(ClientCache.MAX_CACHE_SIZE)
+
+    """
+    lookup entry in cache or request from database
+    """
+    def read(self, read_time, table, key, max_age=None, no_cache=False):
+        if (read_time != None and type(read_time) != TxClock):
+            raise TypeError("read_time must be TxClock")
+
+        # If the value is in cache, just return it.
+        cache_result = self.cache_map.get(table, key, read_time)
+        if (cache_result) != None:
+            return cache_result
+
+        # Otherwise, create the read request
+        response = self.http_facade.read(read_time, table, key, 
+            max_age, no_cache)
+
+        # Evaluate the response
+        if (response.status != 200):
```

Comment:
The status and headers are HTTP things. Translating them into an exception belongs in HTTP facade.